cores/esp8266/core_esp8266_waveform.cpp

```diff
@@ -173,137 +151,133 @@ int ICACHE_RAM_ATTR stopWaveform(uint8_t pin) {
-#endif
-
-
 static ICACHE_RAM_ATTR void timer1Interrupt() {
   // Optimize the NMI inner loop by keeping track of the min and max GPIO that we
   // are generating.  In the common case (1 PWM) these may be the same pin and
   // we can avoid looking at the other pins.
   static int startPin = 0;
   static int endPin = 0;
 
-  uint32_t nextEventCycles = microsecondsToClockCycles(MAXIRQUS);
-  uint32_t timeoutCycle = GetCycleCountIRQ() + microsecondsToClockCycles(14);
+  constexpr uint32_t isrTimeoutCycles = microsecondsToClockCycles(14);
+  const uint32_t isrStart = ESP.getCycleCount();
+  uint32_t nextEventCycle = isrStart + microsecondsToClockCycles(MAXIRQUS);
 
-  if (waveformToEnable || waveformToDisable) {
+  if (waveformsToEnable || waveformsToDisable) {
     // Handle enable/disable requests from main app.
-    waveformEnabled = (waveformEnabled & ~waveformToDisable) | waveformToEnable; // Set the requested waveforms on/off
-    waveformState &= ~waveformToEnable;  // And clear the state of any just started
-    waveformToEnable = 0;
-    waveformToDisable = 0;
+    waveformsEnabled = (waveformsEnabled & ~waveformsToDisable) | waveformsToEnable; // Set the requested waveforms on/off
+    waveformsState &= ~waveformsToEnable;  // And clear the state of any just started
+    waveformsToEnable = 0;
+    waveformsToDisable = 0;
     // Find the first GPIO being generated by checking GCC's find-first-set (returns 1 + the bit of the first 1 in an int32_t)
-    startPin = __builtin_ffs(waveformEnabled) - 1;
+    startPin = __builtin_ffs(waveformsEnabled) - 1;
     // Find the last bit by subtracting off GCC's count-leading-zeros (no offset in this one)
-    endPin = 32 - __builtin_clz(waveformEnabled);
+    endPin = 32 - __builtin_clz(waveformsEnabled);
   }
 
   bool done = false;
-  if (waveformEnabled) {
+  if (waveformsEnabled) {
+    uint32_t now = ESP.getCycleCount();
     do {
-      nextEventCycles = microsecondsToClockCycles(MAXIRQUS);
+      nextEventCycle = now + microsecondsToClockCycles(MAXIRQUS);
       for (int i = startPin; i <= endPin; i++) {
-        uint32_t mask = 1<<i;
+        uint32_t mask = 1UL<<i;
 
         // If it's not on, ignore!
-        if (!(waveformEnabled & mask)) {
+        if (!(waveformsEnabled & mask))
           continue;
-        }
 
-        Waveform *wave = &waveform[i];
-        uint32_t now = GetCycleCountIRQ();
+        Waveform *wave = &waveforms[i];
 
-        // Disable any waveforms that are done
-        if (wave->expiryCycle) {
-          int32_t expiryToGo = wave->expiryCycle - now;
-          if (expiryToGo < 0) {
-              // Done, remove!
-              waveformEnabled &= ~mask;
-              if (i == 16) {
-                GP16O &= ~1;
-              } else {
-                ClearGPIO(mask);
-              }
-              continue;
-            }
+        // Check for toggles etc.
+        if (ExpiryState::UPDATE == wave->hasExpiry) {
```

Comment:
There's a whole lot of new logic for expiry detection here.  What's the intended use case?  The major use seems like it would be the Tone() call, which seems to not warrant so much precision.

Reply:
There were two requirements that had set myself: more precise timing (actually Servo rather than Tone, but it comes down to the same issues).
I also wanted to be able to stop a waveform precisely, and keep the line level at the moment instead of always going LOW.
All this works now - if someone could put an oscilloscope to the new code and check max frequency etc, that would be great.

Reply:
> There were two requirements that had set myself: more precise timing (actually Servo rather than Tone, but it comes down to the same issues).
> I also wanted to be able to stop a waveform precisely, and keep the line level at the moment instead of always going LOW.
> All this works now - if someone could put an oscilloscope to the new code and check max frequency etc, that would be great.

With your current code as of 4 days ago ([last change](https://github.com/dok-net/arduino-esp8266/commit/ceca2051b45068cb544cb4db06a5de7607ecf26e)), here's what I'm seeing:
The phase between 2 pins PWM-controlled jumps by 60us when the duty cycle hits the rails (0 & 1023).  Maybe my little test program below is in error.  Scope is triggered on the rising edge of the top trace.

![PWM_test](https://user-images.githubusercontent.com/53401742/74595089-2ab74800-5003-11ea-8c60-8f60ce6b709b.gif)

Here's a short 1.5meg/8sec video of it hitting the limits and suddenly jumping by 60us: http://www.mediafire.com/file/mnc28ogkzc366or/VID_PWM_%25237022.mp4/file
(sorry for the poor quality, my phone won't focus  at that distance)

```c
const int ch1Pin = 14; 
const int ch2Pin = 2; 

void setup() {
}

void loop() {
  // increase duty cycle, decrease for second channel
  for(int dutyCycle = 0; dutyCycle < 1023; dutyCycle++){   
    analogWrite(ch1Pin, 1023 - dutyCycle);
    analogWrite(ch2Pin, dutyCycle);
    delay(3);
  }

  // decrease duty cycle, increase for second channel
  for(int dutyCycle = 1023; dutyCycle > 0; dutyCycle--){
    analogWrite(ch1Pin, 1023 - dutyCycle);
    analogWrite(ch2Pin, dutyCycle);
    delay(3);
  }
}
```

Reply:
@tech-tx Absolutely fantastic, awesome, thank you so much!
What happens proves my point - and now I should adapt analogWrite to make use of the zero-duration cycle, hold level, enhancement. Right now it calls stopWaveform, so phase can only be lost at 0 and 1023.
Waveform will then have to be stopped by explicit digitalWrite instead of analogWrite(0) or analogWrite(PWMRANGE), to indicate that phase no longer matters.

Are there any known-good min and max frequency values for analogWrite() that you could graph for continuing accuracy just yet?

Reply:
I'm seeing strange results at **all** frequencies, even with the current git.  For a simple test (if you don't have a 'scope), hook up an LED to a port pin and drive it with 20KHz PWM, slowly increasing from 1 to 1023 duty cycle with at least a few ms between duty cycle changes.  It's not a smooth progression.  Most LEDs will be able to handle 20KHz PWM (a few will even hit 200KHz).  The aberrant behavior at higher frequency is pretty obvious on an LED.  

It's less noticeable at 1KHz, but there aren't 1023 steps there, more like 170-180 discrete steps from full-off to full-on.  At 20KHz it's more like 8 to 10 steps, with lots of jitter above 80% duty cycle, and the frequency jumps around above 90% duty cycle.  I think Earle is going to look at it to make sure I haven't made a dumb newbie mistake. ;-)

Reply:
I understand the PWM resolution is 1 microsecond (80 cycles). So to get full resolution you need a complete cycle of 1024 microseconds or 976Hz. That's why in Tasmota we chose 880Hz which is close enough. 

I believe linearity tests should be done at a similar frequency. 

20KHz base pwm frequency means roughly being able to drive at 20MHz which is not realistic with an esp at 80MHz - unless you spend all your CPU time in the PWM loop.

Reply:
@tech-tx when you say current git, is that the master branch or the latest push to this PR? Besides that, I am left a bit clueless how I am not seeing what you described in yesterday's oscilloscope videos. Anyway, is the phase loss part fixed now?

Reply:
Current git, not your PR.  I'll start doing a more exhaustive look at yours tomorrow.  I had a weird error here, and thought the current git was all screwy.  Git said I was even with master, but it wouldn't work right, so I mistrusted anything I'd seen on yours.  I eventually tried current git on my main desktop and it turns out the laptop had one or more corrupt files that git didn't notice.  Wipe, reinstall, re-pull, and it's back again.  I hate days like that...

Reply:
Sorry, I was having problems and then Real Life rudely intruded.  Looking at your latest code (I just pulled it this morning, Feb 29) I see significant jitter if edges on other pins are walking past the servo edges.  Currently I'm running 2 PWM channels that are async to the servo, and one tone output (different async rate), as well as 2 servo channels.  I only have one servo that runs at 3.3V, so it gets moved around a little.  When you see the rapid changes around 7 seconds in, I can hear the servo trying to comply with what looks like a 4 degree change.
It's easily possible I've futzed something up.  Here's the video:
https://youtu.be/1Ct2ZF_CRPk
Here's the two outer markers shown, since the video is blurry:
![waveform PR 29-Feb-2020 thumbnail](https://user-images.githubusercontent.com/53401742/75614630-61f81f80-5b00-11ea-9c6a-4e1e8cf9be6d.jpg)
For comparison, I see around 6.5us total jitter with the current master.
I've checked that my master code is correct, but grabbing a copy of your PR files may be iffy.

Reply:
Next issue looks like a math error on setting the next edge.  The current master works in multiples of 1us step size for the PWM duty cycle.  That means the fastest PWM that has all 1022 steps will be 977Hz; at 978Hz it drops the 1/1022 step and the signal goes away.  At slower PWM frequency the step size changes from 1us to 2us, and in between that it misses up to half of the PWM steps before it shifts to the next longer / shorter 1us multiple.

Your PR on the other hand is running 7.76us step sizes at 977Hz, resulting in about 1/8th the total number of steps.  You can actually see it if you hook up an LED to the PWM pin, set the pause between duty cycle changes to 20ms, and then watch closely as the duty cycle approaches minimum or maximum, depending on whether you ground the other leg or put it at 3.3V.  As the LED is going very dim to nearly off, you can see the last 5 or 6 levels before it goes to minimum brightness.  At higher levels you can't perceive the change. With the current master the step size is smaller, and there's no visual step-wise drop in brightness.  Here's the screen caps from the scope showing the difference (all settings were identical):
![master 977Hz PWM step sizes at 977Hz](https://user-images.githubusercontent.com/53401742/75616251-6e867300-5b14-11ea-9ba2-3850b8a0600e.gif)
![waveform PR PWM step sizes at 977Hz](https://user-images.githubusercontent.com/53401742/75616254-77774480-5b14-11ea-899c-0ca336aa83c3.gif)
![waveform PR PWM zoomed on step sizes](https://user-images.githubusercontent.com/53401742/75616256-7fcf7f80-5b14-11ea-87ec-2f32e1c042fb.gif)
That last one is zoomed in to get a more accurate number on the PWM step size.

The final capture is from a logic analyzer.  The top 2 traces are PWM channels with the decoded PWM duty cycle below each one.  The bottom trace toggles each time the duty cycle is changed.  You can see that the actual duty cycle is only changing about every 8 increments, which is that 7.6us step size.
![waveform PR 977Hz 1ms pause](https://user-images.githubusercontent.com/53401742/75616436-4ef04a00-5b16-11ea-8451-c35746cec300.png)